### PR TITLE
fix(controller/ui): fix pod with sidecar state

### LINKED ISSUE #18725: Containers in pods with native sidecars don't show as "ready"
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
In the tree view, pods with one or more [native sidecars](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) show all containers as not ready i.e. 0/2, even if they are. 

**To Reproduce**
Define a Deployment with at least one container set up as a [native sidecar](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) and sync. When synced none of the containers in any of the resulting pods will appear ready.

**Expected behavior**
Pods in the tree view would show that containers are ready if they are.

**Screenshots**

Argocd
image

Kubectl
image

**Version**

```shell
v2.11.3+3f344d5
```

**Logs**

```
Paste any relevant application logs here.
```

### PR DESCRIPTION
**Pull request by linghaoSu**
Fixes https://github.com/argoproj/argo-cd/issues/18725

## Before:

image
image

## After

image
image

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by alexef** [comment]
@linghaoSu why is this still a draft? anything missing?

**Comment by alexef** [comment]
I assume you also want to test in https://github.com/argoproj/argo-cd/blob/1e359c8c0d25dc3a223ab184c1d623bb4d68aa5a/controller/cache/info_test.go? Let me know if you have time to continue with this PR, otherwise I can also try to take over.

We really need this fix to land.

**Comment by linghaoSu** [comment]
Basic functionality is complete and will be added for testing later today

References the dashboard pr: https://github.com/kubernetes/dashboard/pull/9483/files#diff-207a2dad9e596869b3612d332964328e081f20fb9cbeb6f9bbb41dfa13472152R39-R124

File: controller/cache/info.go [hunk 1]
```diff
@@ -278,6 +278,33 @@ func populateIstioVirtualServiceInfo(un *unstructured.Unstructured, res *Resourc
...
+		return false
+	}
+	if initContainer.RestartPolicy == nil {
+		return false
+	}
+
+	return *initContainer.RestartPolicy == v1.ContainerRestartPolicyAlways
+}
+
+// IsPodPhaseTerminal returns true if the pod's phase is terminal.
```
Comment (alexef) [inline controller/cache/info.go:303]: comment not needed

File: controller/cache/info.go [hunk 2]
```diff
@@ -370,7 +412,7 @@ func populatePodInfo(un *unstructured.Unstructured, res *ResourceInfo) {
 	// and https://github.com/kubernetes/kubernetes/issues/90358#issuecomment-617859364
 	if pod.DeletionTimestamp != nil && pod.Status.Reason == "NodeLost" {
 		reason = "Unknown"
-	} else if pod.DeletionTimestamp != nil {
+	} else if pod.DeletionTimestamp != nil && !IsPodPhaseTerminal(podPhase) {
```
Comment (alexef) [inline controller/cache/info.go:415]: why would podDeletionTimestamp not be nil, but IsPodPhaseTerminal be true?

File: ui/src/app/applications/components/utils.tsx [hunk 3]
```diff
@@ -1053,7 +1085,7 @@ export function getPodStateReason(pod: appModels.State): {message: string; reaso
     if ((pod as any).metadata.deletionTimestamp && pod.status.reason === 'NodeLost') {
         reason = 'Unknown';
         message = '';
-    } else if ((pod as any).metadata.deletionTimestamp) {
+    } else if ((pod as any).metadata.deletionTimestamp && !isPodPhaseTerminal(podPhase)) {
```
Comment (alexef) [inline ui/src/app/applications/components/utils.tsx:1088]: so the cases we exclude here are Pending, Running and Unknown

File: controller/cache/info_test.go [hunk 4]
```diff
@@ -285,6 +285,116 @@ func TestGetPodInfo(t *testing.T) {
...
+        terminated: 
+          exitCode: 0
+          reason: "Completed"
+    phase: "Running"
+`)
+
+	info := &ResourceInfo{}
+	populateNodeInfo(pod, info, []string{})
+	assert.Equal(t, []v1alpha1.InfoItem{
+		{Name: "Status Reason", Value: "Running"},
```
Comment (alexef) [inline controller/cache/info_test.go:337]: maybe add one more test for the clasic use case of init 0/1 while the regular initcontainer didn't finish yet?

**Review by alexef — APPROVED** [review]
couple of questions

File: controller/cache/info.go (hunk 2 above)
Comment (linghaoSu) [reply controller/cache/info.go:415]: Thanks for your review!

I found this [issue](https://github.com/kubernetes/kubectl/issues/1595), and the dashboard PR could potentially reference this [change](https://github.com/kubernetes/kubernetes/pull/122038/files#diff-8083c7028e9b5553d36b3edd6d4a533c7a20e8ba883990de84b061d728572cd0R993).

So when a pod is in a Succeeded or Failed state, but has a finalizer and deletionTimestamp, based on this issue, we should prevent it from being displayed as "Terminating."

Comment (alexef) [reply controller/cache/info.go:415]: makes sense, thank you for the clarification. maybe add a comment?

Comment (linghaoSu) [reply controller/cache/info.go:422]: added

File: controller/cache/info.go (hunk 1 above)
Comment (linghaoSu) [reply controller/cache/info.go]: removed

**Comment by linghaoSu** [comment]
I included restartable initContainers in the total container calculation, following the approach used in kubectl.

https://github.com/kubernetes/kubernetes/blob/f7fef0384ec1c666e6b68cae83a3b3c6d95aa7b0/pkg/printers/internalversion/printers.go#L887-L892

@alexef If it's convenient, could you please review it again?

**Review by alexef — APPROVED** [review]
great. now we only need an approver :)

**Comment by linghaoSu** [comment]
Hi @crenshaw-dev,

I believe this PR is ready for review. Could you please take a look? This [issue](https://github.com/argoproj/argo-cd/issues/18725) has 16 thumbs up, so it might be helpful to address it.

If you spot any potential issues with this PR, please let me know, and I'll be happy to work on fixing them.

Thanks!

File: controller/cache/info.go [hunk 5]
```diff
@@ -296,6 +296,32 @@ func populateIstioServiceEntryInfo(un *unstructured.Unstructured, res *ResourceI
...
+		return false
+	}
+	if initContainer.RestartPolicy == nil {
+		return false
+	}
+
+	return *initContainer.RestartPolicy == v1.ContainerRestartPolicyAlways
+}
+
+func IsPodPhaseTerminal(phase v1.PodPhase) bool {
```
Comment (crenshaw-dev) [inline controller/cache/info.go:321]: Can we make this function private?

File: controller/cache/info_test.go [hunk 6]
```diff
@@ -308,6 +308,374 @@ func TestGetPodInfo(t *testing.T) {
...
+        state:
+          waiting: {}
+        started: false
+    containerStatuses:
+      - ready: false
+        restartCount: 0
+        state:
+          waiting: {}
+    conditions:
+      - type: PodInitialized
```
Comment (crenshaw-dev) [inline controller/cache/info_test.go:516]:
```suggestion
      - type: ContainersReady
        status: "False"
      - type: Initialized
```

I think this should be `Initialized`. Pulled a sample from another pod:

```yaml
status:
  conditions:
    - lastProbeTime: null
      lastTransitionTime: '2024-09-22T10:48:10Z'
      status: 'True'
      type: Initialized
```

Also added another condition just to exercise the part of `isPodInitializedConditionTrue` that skips irrelevant conditions.

File: controller/cache/info_test.go [hunk 7]
```diff
@@ -308,6 +308,374 @@ func TestGetPodInfo(t *testing.T) {
...
+        state:
+          running: {}
+        started: false
+    containerStatuses:
+      - ready: false
+        restartCount: 0
+        state:
+          waiting: {}
+    conditions:
+      - type: PodInitialized
```
Comment (crenshaw-dev) [inline controller/cache/info_test.go:569]:
```suggestion
      - type: ContainersReady
        status: "False"
      - type: Initialized
```

File: controller/cache/info_test.go [hunk 8]
```diff
@@ -308,6 +308,374 @@ func TestGetPodInfo(t *testing.T) {
...
+    containerStatuses:
+      - ready: true
+        restartCount: 4
+        state:
+          running: {}
+        lastTerminationState:
+          terminated:
+            finishedAt: "2023-10-01T00:00:00Z" # Replace with actual time
+    conditions:
+      - type: PodInitialized
```
Comment (crenshaw-dev) [inline controller/cache/info_test.go:625]:
```suggestion
      - type: ContainersReady
        status: "False"
      - type: Initialized
```

File: controller/cache/info.go [hunk 9]
```diff
@@ -388,7 +435,9 @@ func populatePodInfo(un *unstructured.Unstructured, res *ResourceInfo) {
 	// and https://github.com/kubernetes/kubernetes/issues/90358#issuecomment-617859364
 	if pod.DeletionTimestamp != nil && pod.Status.Reason == "NodeLost" {
 		reason = "Unknown"
-	} else if pod.DeletionTimestamp != nil {
+		// If the pod is being deleted and the pod phase is not succeeded or failed, set the reason to "Terminating".
+		// See https://github.com/kubernetes/kubectl/issues/1595#issuecomment-2080001023
+	} else if pod.DeletionTimestamp != nil && !IsPodPhaseTerminal(podPhase) {
 		reason = "Terminating"
```
Comment (crenshaw-dev) [inline controller/cache/info.go:441]: Looks like the tests aren't hitting this. Needs another test case?

File: controller/cache/info.go [hunk 10]
```diff
@@ -324,13 +351,34 @@ func populatePodInfo(un *unstructured.Unstructured, res *ResourceInfo) {
 		res.Images = append(res.Images, image)
 	}
 
+	// If the Pod carries {type:PodScheduled, reason:SchedulingGated}, set reason to 'SchedulingGated'.
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == v1.PodScheduled && condition.Reason == v1.PodReasonSchedulingGated {
+			reason = v1.PodReasonSchedulingGated
+		}
```
Comment (crenshaw-dev) [inline controller/cache/info.go:358]: I don't see a test hitting this. Need to add one?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Thanks for the PR! Just a few nitpicks and requests for code coverage.

File: ui/src/app/applications/components/utils.tsx [hunk 11]
```diff
@@ -987,23 +987,55 @@ export const OperationState = ({app, quiet}: {app: appModels.Application; quiet?
     );
 };
 
+function isPodInitializedConditionTrue(status: any): boolean {
+    for (const condition of status.conditions) {
```
Comment (crenshaw-dev) [inline ui/src/app/applications/components/utils.tsx:991]:
```suggestion
    for (const condition of status.conditions) {
        if (!status.conditions) {
            return false;
        }
```

I'm not sure how Babel will transpile this down, but in node I got a runtime error with a missing conditions field.

Comment (linghaoSu) [reply ui/src/app/applications/components/utils.tsx:995]: I had overlooked the fact that a field might be missing in some cases. I've now moved the existence check logic before the loop.

File: controller/cache/info.go (hunk 10 above)
Comment (linghaoSu) [reply controller/cache/info.go:358]: add test case fot this line

File: controller/cache/info.go (hunk 9 above)
Comment (linghaoSu) [reply controller/cache/info.go:441]: I've added more test cases for this.

File: controller/cache/info.go (hunk 5 above)
Comment (linghaoSu) [reply controller/cache/info.go]: Fixed.

**Comment by linghaoSu** [comment]
Hi @crenshaw-dev,

Thank you for your suggestions! I believe I've addressed all the comments. Could you please take another look?

**Comment by crenshaw-dev** [comment]
Looking good! Bear with me a bit, I want to scope-creep this PR a bit to solve some problems with the UI code:

image

Notice the `Init: 1)` message. I think we can fix that pretty easily. I'll open a PR against yours.

**Comment by crenshaw-dev** [comment]
https://github.com/linghaoSu/argo-cd/pull/1

**Review by crenshaw-dev — APPROVED** [review]
This LGTM, will follow up with the UI fix. Thanks!

**Comment by crenshaw-dev** [comment]
@linghaoSu would you mind opening cherry-picks for 2.10-2.13?

**Comment by linghaoSu** [comment]
Hi @crenshaw-dev,

I've manually cherry-picked this fix to 2.10-2.13. Could you please take a look at those PRs?

**Comment by linghaoSu** [comment]
@crenshaw-dev , However in 2.10 and 2.11, RestartPolicy is not a valid field in InitContainer, so this would cause a lint error. Perhaps we shouldn't cherry-pick this fix to those two releases.

image

**Pull request merged**